src/test/java/io/nats/client/MsgDeliveryPoolTest.java

```diff
@@ -23,8 +25,82 @@
 public class MsgDeliveryPoolTest extends BaseUnitTest {
 
     @Test
+    public void testMsgDeliveryWorker() throws Exception {
+        final MsgDeliveryWorker worker = new MsgDeliveryWorker();
+        assertTrue(worker.getName().contains("delivery"));
+        worker.start();
+        final CountDownLatch latch = new CountDownLatch(1);
+        AsyncSubscriptionImpl sub = new AsyncSubscriptionImpl(null, "foo", null, new MessageHandler() {
+            public void onMessage(Message msg) {
+                latch.countDown();
+            }
+        });
+        sub.lock();
+        sub.setDeliveryWorker(worker);
+        sub.unlock();
+        worker.lock();
+        worker.postMsg(new Message("hello".getBytes(), "foo", null, sub));
+        worker.unlock();
+        latch.await();
+        worker.shutdown();
+    }
+
+    @Test
+    public void testMsgDeliveryWorkerShutdownFromCB() throws Exception {
+        final MsgDeliveryWorker worker = new MsgDeliveryWorker();
+        assertTrue(worker.getName().contains("delivery"));
+        worker.start();
+        final CountDownLatch latch = new CountDownLatch(1);
+        AsyncSubscriptionImpl sub = new AsyncSubscriptionImpl(null, "foo", null, new MessageHandler() {
+            public void onMessage(Message msg) {
+                latch.countDown();
+                worker.shutdown();
+            }
+        });
+        sub.lock();
+        sub.setDeliveryWorker(worker);
+        sub.unlock();
+        worker.lock();
+        worker.postMsg(new Message("hello".getBytes(), "foo", null, sub));
+        worker.unlock();
+        latch.await();
+        // Double shutdown is fine
+        worker.shutdown();
+    }
+
+    @Test
+    public void testMsgDeliveryWorkerExceptionInCB() throws Exception {
+        final MsgDeliveryWorker worker = new MsgDeliveryWorker();        
+        worker.start();
+        final CountDownLatch latch = new CountDownLatch(2);
+        AsyncSubscriptionImpl sub = new AsyncSubscriptionImpl(null, "foo", null, new MessageHandler() {
+            public void onMessage(Message msg) {
+                latch.countDown();
+                throw new RuntimeException("On purpose");
+            }
+        });
+        sub.lock();
+        sub.setDeliveryWorker(worker);
+        sub.unlock();
+        // Post 2 messages
+        worker.lock();
+        worker.postMsg(new Message("hello".getBytes(), "foo", null, sub));
+        worker.postMsg(new Message("hello".getBytes(), "foo", null, sub));
+        worker.unlock();
+        // Callback should be invoked twice, even if throwing exception
+        latch.await(2, TimeUnit.SECONDS);
+        worker.shutdown();
+    }
+
+    @Test
     public void testMsgDeliveryPool() throws Exception {
         MsgDeliveryPool pool = new MsgDeliveryPool();
+        // Check zero or negative size is ignored.
+        pool.setSize(-2);
```

Comment:
Should this throw an exception?

Reply:
Other than this, LGTM!

Reply:
The user facing API is `Nats.setMsgDeliveryThreadPoolSize()` which throws an `IllegalArgumentException` if size is `<=0`. So the low level `MsgDeliveryPool.setSize()` does not need to - since it is assumed to be used correctly. The unit test calls `setSize(-2)` and verifies that size is still `0`.
I am adding in javadoc for Nats API that it can throw exception if size is <= 0.